zcash_client_backend/src/data_api.rs

```diff
@@ -1468,6 +1471,72 @@ pub trait WalletWrite: WalletRead {
         birthday: &AccountBirthday,
     ) -> Result<(Self::AccountId, UnifiedSpendingKey), Self::Error>;
 
+    /// Tells the wallet to track a specific account index for a given seed.
+    ///
+    /// Returns the account for the newly-created wallet database entry, along with the
+    /// associated [`UnifiedSpendingKey`]. Note that the unique account identifier should *not* be
+    /// assumed equivalent to the ZIP 32 account index. It is an opaque identifier for a pool of
+    /// funds or set of outputs controlled by a single spending authority.
+    ///
+    /// The ZIP-32 account index may be obtained by calling [`WalletRead::get_account`]
+    /// with the returned account identifier.
+    ///
+    /// If `birthday.height()` is below the current chain tip, this operation will
+    /// trigger a re-scan of the blocks at and above the provided height. The birthday height is
+    /// defined as the minimum block height that will be scanned for funds belonging to the wallet.
+    ///
+    /// For new wallets, callers should construct the [`AccountBirthday`] using
+    /// [`AccountBirthday::from_treestate`] for the block at height `chain_tip_height - 100`.
+    /// Setting the birthday height to a tree state below the pruning depth ensures that reorgs
+    /// cannot cause funds intended for the wallet to be missed; otherwise, if the chain tip height
+    /// were used for the wallet birthday, a transaction targeted at a height greater than the
+    /// chain tip could be mined at a height below that tip as part of a reorg.
+    ///
+    /// If `seed` was imported from a backup and this method is being used to restore a previous
+    /// wallet state, you should use this method to add all of the desired accounts before scanning
+    /// the chain from the seed's birthday height.
+    ///
+    /// By convention, wallets should only allow a new account to be generated after confirmed
+    /// funds have been received by the currently-available account (in order to enable automated
+    /// account recovery).
+    ///
+    /// Panics if the length of the seed is not between 32 and 252 bytes inclusive.
+    ///
+    /// [ZIP 316]: https://zips.z.cash/zip-0316
+    fn import_account_hd(
+        &mut self,
+        seed: &SecretVec<u8>,
+        account_index: zip32::AccountId,
+        birthday: &AccountBirthday,
+    ) -> Result<(Self::Account, UnifiedSpendingKey), Self::Error>;
```

Comment:
While I think that UFVK import is a necessary feature, I'm much less convinced about multi-seed support. The whole idea of a seed phrase is that there's a single piece of information that is needed in order to provide recoverability for the funds in a wallet. On the other hand, I guess that this can be useful for recovery scenarios, but I really don't care much for the idea of a wallet having to manage multiple spend authority roots.

Reply:
Is `AccountCollision` intended for the case where `create_account` attempts to create an account for a seed and zip32 account ID that were previously created via `import_account_hd`?

Reply:
I don't think `import_account_id` introduced multi-seed support. We did that a while back when seed fingerprinting was added. 
The new thing that this method adds is direct control over the account index. It doesn't have to be a blind +1 over whatever the highest index was before. The app can actually be aware of and control it. And while users will typically be blissfully unaware of it, from an API level I strongly believe this should be exposed. 
In my .NET library, I expose a `ZcashAccount` type that you initialize with a seed and index (or a viewing key) and you can pass that around. When it's time to hook it up to a lightwallet to start syncing, I just need to be able to get that account created, even if the index is 5 and 4 other accounts weren't created first.

`AccountCollision` is for a couple cases: a colliding account index, or importing an account by viewing key that happens to collide with pre-existing account (which might be another viewing key imported account or an HD derived account).

Reply:
And I have my own personal use case for explicit account indexes: for my seed phrase, account indexes 0-2 are used and _very_ old. I don't want to sync them _at all_ going forward. I want to keep my seed phrase by using index 3 for a new account. Given a new wallet, I can see myself using an account import feature, specifying my seed phrase and account index 3 as the _starting_ index, with a much higher birthday height. No reason to sync 0-2 at all.

Reply:
I think that adding an explicit `sunset_height` to the `accounts` table would be a good way to serve this. Sunsetted accounts won't have their keys used for scanning.

Reply:
Yes, if that feature existed, that might work. Still, IMO limiting the API in this way is tying people's hands in order to try to force a particular wallet UX. I see account management as something that the app (or a higher level library as is in my case) to be something that can and should occur separately from the engine that does sync. My mental model is that I jump across the FFI layer to sync the blockchain and create transactions. The fact that it tracks accounts is only nice if it _helps_ me. 
But I guess I can always just use the UFVK import function and bypass seeds and account indexes at this layer entirely so that I can manage them at my preferred level. So if you feel strongly against this API I'll remove it.

Reply:
After some additional thought, I'm okay with this API. Also, I agree that returning `Self::Account` instead of `Self::AccountId` is a nicer API, and would support changing the existing method (in a separate PR).